Better support for dnf5 in Fedora 41/42

gregoster opened the pull request:
'grouplist' and 'groupinfo' are each now two words. 'group' and 'group-id' have been replaced with 'name' and 'id', so check for those too.

### What does this PR do?
Improve support for installing via groups with dnf5 on Fedora 41/42.  Issue found when attempting to run salt 3007 on Fedora 42.  Problem also exists in master/main, so starting there.

### What issues does this PR fix or reference?
Fixes salt failing on a Fedora 41/42 install when (e.g.) installing KDE Fonts via a group:
```
Fonts:
  pkg.group_installed
```
The name of the group doesn't seem to matter -- all groups fail, as 'groupinfo' and 'grouplist' have been deprecated since dnf5 as shipped with Fedora 41.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the test documentation for details on how to implement tests
into Salt's test suite:
https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html -->
- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
No

<!-- Please review Salt's Contributing Guide for best practices and guidance in
choosing the right branch:
https://docs.saltproject.io/en/master/topics/development/contributing.html -->

<!-- Additional guidance for pull requests can be found here:
https://docs.saltproject.io/en/master/topics/development/pull_requests.html -->

<!-- See GitHub's page on GPG signing for more information about signing commits
with GPG:
https://help.github.com/articles/signing-commits-using-gpg/ -->

Commit:
Better support for dnf5 in Fedora 41/42

'grouplist' and 'groupinfo' are each now two words.
'group' and 'group-id' have been replaced with 'name' and 'id',
so check for those too.

twangboy reviewed (changes requested):
Please create a changelog and write a test for this.

Commit:
Add a changelog entry for 67975.

twangboy reviewed (changes requested):
Could you write a simple test here, or ensure that existing tests cover these changes

gregoster commented:
Please see my previous comment.  I've tried to build a test, but I'm failing right now to even get a basic test environment setup.  
```Command python -m pip install --progress-bar=off -r requirements/static/ci/py3.13/linux.txt failed with exit code 1:  ... <many lines of failure> ...```
So while I could probably dig enough to figure out what to write for a test case, I have no way to test my test :(

twangboy commented:
```
python -m venv venv
source venv/bin/activate
pip install nox==2022.8.7
nox -e "pytest-zeromq-3(coverage=False)" -- -vvv --run-destructive --core-tests --run-slow  .\tests\pytests\unit\modules\test_yumpkg.py
```

Commit:
Correctly parse the dnf5 group info format.

The dnf5 'group info' output differs from the dnf 'groupinfo' output.
In particular, a package name is now also listed on the first line of
each section.  Need to also strip leading spaces from lines.

Commit:
Write a test for dnf5 'group info' parsing for PR 67975.

gregoster commented:
Test added for the dnf5 'group info' bits.  

The test revealed that the parsing of dnf5 'group info' output was also broken, so added a fix for that too.

Not sure how to test the "groupinfo" -> "group", "info" and "grouplist" -> "group", "list" changes though.

Commit:
The 'hidden' option requires two leading dashes for dnf5.

gregoster commented:
Ahh.. thanks... I now see there is a _yum() function which should be sufficient for distinguishing whether it's running as yum/dnf/dnf5.

Commit:
Add dnf5 group_list test, and limit 67975 tests to dnf5.

dnf5 group_list-specific test was missing, but these tests
should only be run against dnf5, as yum/dnf provide different
output for 'grouplist' and 'groupinfo'.

Commit:
Make group_list() parse dnf5 'group list' output.

dnf5 'group list' output needs to be handed differently
than yum/dnf 'grouplist' output.  Also condition on
_yum() so we don't break 'grouplist' and 'groupinfo' on
systems using yum/dnf.

Commit:
Fix minor comma formatting.

twangboy commented:
A few more pre-commit and lint failures here

Commit:
Whitespace/formatting police.

gregoster commented:
> A few more pre-commit and lint failures here

Thanks.  I'd seen these sorts of errors months ago, but didn't realize what they were about.   Should be closer now.

Commit:
More formatting/whitespace police.

Commit:
Hopefully last whitespace/formatting fix.

Commit:
If "name" is in the group info, then it's a package group.  Preserve
the case of the package name when adding it to the list.

twangboy reviewed (changes requested):
Please create a changelog and a test

Commit:
Fix dnf5 group_info test and clean up string-membership checks

- test_67975_dnf5_group_info: add missing "type": "package group" key to
  the expected dict (the new elif "name" in g_info branch populates it).
- yumpkg group_list/group_info: replace `_yum() in ("dnf5")`,
  `pkg_id not in ("id")`, and `pkg_installed in ("yes")` with `==`/`!=`
  comparisons. These were substring-in-string checks (no trailing comma
  on a single-string tuple); they worked for exact equality but would
  have accepted any single character in those strings as a match.
- yumpkg group_info: drop stray space inside `re.match( r"..." )` so
  black is happy.

Co-authored-by: Greg Oster <oster@fween.ca>

ggiesen commented:
Heads-up that this also affects the 3006 LTS line: dnf5 ships on Fedora 41+ and AlmaLinux/RHEL 10, and on 3006.26 `pkg.group_list` / `pkg.group_info` (and therefore `pkg.group_installed`) already fail there -- confirmed on Fedora 44:

```
Unknown argument "grouplist" for command "dnf5"
```

Since 3006.x is supported into 2027 and the contributing guide has bug fixes start on the oldest supported branch and merge forward, would it make sense to target this at `3006.x` rather than `master`? The group code has diverged a little between the two (3006.x carries the #60276 env-group member-expansion change), so it'd be a small rebase rather than a straight base swap.

If it helps, I'm happy to put up a 3006.x backport that preserves your authorship and adds the changelog + tests -- just say the word, or feel free to retarget this one. Thanks for chasing this down.

gregoster commented:
> If it helps, I'm happy to put up a 3006.x backport that preserves your authorship and adds the changelog + tests -- just say the word, or feel free to retarget this one. Thanks for chasing this down.

I'm happy to have you (or anyone else) take over the lead on this :)

twangboy commented on salt/modules/yumpkg.py:
```diff
@@ -2512,6 +2512,28 @@ def group_list():
+            line_lc = line.lower()
+            # split line into 3 parts: ID (no spaces), Name (contains spaces), and
+            # Installed (one of 'yes' or 'no')
+            match = re.match(r"^(\S+?)\s+(.+?)\s*(yes|no)$", line_lc)
```
The Problem
The current regex ``r"^(\S+?)\s+(.+?)\s*(yes|no)$"`` relies on a lazy match (``.+?``) for the group name. If a group name naturally contains the word "yes" or "no" (e.g., ``"Just (testing) yes"``), the lazy match can cut off early or misalign the columns.

The Fix
Since terminal tables from tools like ``dnf5`` use predictable, whitespace-aligned columns, it is safer to tokenize the line from the right side first to isolate the status, or use a regex that explicitly binds the trailing status to the end of the line while allowing spaces in the name.

```python
# Instead of: match = re.match(r"^(\S+?)\s+(.+?)\s*(yes|no)$", line_lc)

# A more robust regex approach:
# This ensures the group ID is a single word, the status is strictly the last word,
# and everything in between is captured cleanly as the group name.
match = re.match(r"^(\S+)\s+(.+)\s+(yes|no)$", line_lc)
```

Alternatively, you can split by whitespace and pop the trailing status directly, which completely avoids regex backtracking traps:

```python
parts = line.strip().split()
if len(parts) >= 3:
    pkg_installed = parts[-1].lower()
    if pkg_installed in ("yes", "no"):
        pkg_id = parts[0]
        pkg_name = " ".join(parts[1:-1])
        # Proceed with logic...
```

twangboy commented on salt/modules/yumpkg.py:
```diff
@@ -2656,6 +2688,29 @@ def group_info(name, expand=False, ignore_groups=None, **kwargs):
+                        # The difference here from dnf (above) is that this line
+                        # also contains the first package of this section.
+                        # Have to pull out the package name, but not changing the case
+                        rematch_dnf5 = re.match(r"^.*:\s*(.*?)$", line)
```
The Problem
Mutating the ``line`` variable mid-stream using a broad ``re.match(r"^.*:\s*(.*?)$", line)`` can result in an empty string if a header layout has trailing spaces or localized structural content. If ``line`` becomes empty, it can bypass downstream assertions or drop packages.

The Fix
Instead of overwriting the loop context variable ``line`` directly and executing an immediate ``continue``, explicitly extract the package payload into a dedicated variable name and process it safely inside a structured block:

```python
# Instead of mutating the loop variable:
# if rematch_dnf5:
#     line = rematch_dnf5.group(1)

# Safer approach:
match_dnf5 = re.match(
    pkgtypes_capturegroup + r" (?:groups|packages)\s*:\s*(.*?)$",
    line.lower(),
)
if match_dnf5:
    if target_found:
        break
    
    if match_dnf5.group(1) == pkgtype:
        target_found = True
        
        # Safely extract the inline package without losing the loop context
        header_package_match = re.match(r"^[^:]+:\s*(.+)$", line)
        if header_package_match:
            inline_pkg = header_package_match.group(1).strip()
            if inline_pkg and inline_pkg not in completed_groups:
                # Append the inline package immediately here
                if expand and ret["type"] == "environment group":
                    # handle environment group nesting...
                    pass
                else:
                    ret[pkgtype].append(inline_pkg)
        continue
```

twangboy commented on salt/modules/yumpkg.py:
```diff
@@ -2643,7 +2674,8 @@ def group_info(name, expand=False, ignore_groups=None, **kwargs):
         target_found = False
         for line in salt.utils.itertools.split(out, "\n"):
-            line = line.strip().lstrip(string.punctuation)
+            line = line.strip().lstrip(string.punctuation).lstrip()
```
While the second .lstrip() was likely added to catch any new leading spaces exposed after removing punctuation, this structure leaves a silent bug if a line contains trailing punctuation.

To guarantee that the string is completely tightened on both ends regardless of what characters or whitespace get shifted around by the punctuation drop, the final call should be a full .strip().

```python
line = line.strip().lstrip(string.punctuation).strip()
```

ggiesen commented:
@gregoster thanks for the go-ahead. I've put the 3006.x backport up as #69813, with you as co-author.

Since 3006.x's `group_info` already carries the #60276 multi-word member-group `@`-fallback (which master doesn't have yet), I grafted your dnf5 support onto the current 3006.x functions rather than cherry-picking, so that fix is preserved.

@twangboy I also folded in your three review points there:
- `group_list` now tokenizes the row (ID + the trailing `yes`/`no`) instead of the lazy regex, so a name that contains or ends in "yes"/"no" isn't mistaken for the status column -- and as a side effect it no longer drops a row with trailing whitespace (which the regex did).
- `group_info` pulls the dnf5 inline first package into a dedicated variable instead of overwriting `line`.
- The section-line cleaning is a full `.strip()` on both ends.

Validated end to end against live dnf5 5.4.2.1 on Fedora 44: `group_list` parses all 157 groups and `group_info` drops no packages (41 parsed == 41 raw package lines for `development-tools`), and it reproduces your `test_67975_dnf5_group_info` expected output exactly, @gregoster.

This master PR still stands on its own -- happy to help forward-port the same three fixes here if that's useful.

twangboy commented:
Closing in favor of: https://github.com/saltstack/salt/pull/69813
Those fixes will be merged forward in to later branches and will eventually end up on master.
